Mathlib/Order/GaloisConnection.lean

```diff
@@ -0,0 +1,926 @@
+  gi.dual.l_supₛ_u_image _
+#align galois_coinsertion.u_Inf_l_image GaloisCoinsertion.u_infₛ_l_image
+
+theorem u_sup_l [SemilatticeSup α] [SemilatticeSup β] (gi : GaloisCoinsertion l u) (a b : α) :
+    u (l a ⊔ l b) = a ⊔ b :=
+  gi.dual.l_inf_u _ _
+#align galois_coinsertion.u_sup_l GaloisCoinsertion.u_sup_l
+
+theorem u_supᵢ_l [CompleteLattice α] [CompleteLattice β] (gi : GaloisCoinsertion l u) {ι : Sort x}
+    (f : ι → α) : u (⨆ i, l (f i)) = ⨆ i, f i :=
+  gi.dual.l_infᵢ_u _
+#align galois_coinsertion.u_supr_l GaloisCoinsertion.u_supᵢ_l
+
+theorem u_bsupᵢ_l [CompleteLattice α] [CompleteLattice β] (gi : GaloisCoinsertion l u) {ι : Sort x}
+    {p : ι → Prop} (f : ∀ (i) (_ : p i), α) : u (⨆ (i) (hi), l (f i hi)) = ⨆ (i) (hi), f i hi :=
+  gi.dual.l_binfᵢ_u _
+#align galois_coinsertion.u_bsupr_l GaloisCoinsertion.u_bsupᵢ_l
+
+theorem u_supₛ_l_image [CompleteLattice α] [CompleteLattice β] (gi : GaloisCoinsertion l u)
+    (s : Set α) : u (supₛ (l '' s)) = supₛ s :=
+  gi.dual.l_infₛ_u_image _
+#align galois_coinsertion.u_Sup_l_image GaloisCoinsertion.u_supₛ_l_image
+
+theorem u_supᵢ_of_lu_eq_self [CompleteLattice α] [CompleteLattice β] (gi : GaloisCoinsertion l u)
+    {ι : Sort x} (f : ι → β) (hf : ∀ i, l (u (f i)) = f i) : u (⨆ i, f i) = ⨆ i, u (f i) :=
+  gi.dual.l_infᵢ_of_ul_eq_self _ hf
+#align galois_coinsertion.u_supr_of_lu_eq_self GaloisCoinsertion.u_supᵢ_of_lu_eq_self
+
+theorem u_bsupᵢ_of_lu_eq_self [CompleteLattice α] [CompleteLattice β] (gi : GaloisCoinsertion l u)
+    {ι : Sort x} {p : ι → Prop} (f : ∀ (i) (_ : p i), β) (hf : ∀ i hi, l (u (f i hi)) = f i hi) :
+    u (⨆ (i) (hi), f i hi) = ⨆ (i) (hi), u (f i hi) :=
+  gi.dual.l_binfᵢ_of_ul_eq_self _ hf
+#align galois_coinsertion.u_bsupr_of_lu_eq_self GaloisCoinsertion.u_bsupᵢ_of_lu_eq_self
+
+theorem l_le_l_iff [Preorder α] [Preorder β] (gi : GaloisCoinsertion l u) {a b} :
+    l a ≤ l b ↔ a ≤ b :=
+  gi.dual.u_le_u_iff
+#align galois_coinsertion.l_le_l_iff GaloisCoinsertion.l_le_l_iff
+
+theorem strict_mono_l [Preorder α] [Preorder β] (gi : GaloisCoinsertion l u) : StrictMono l :=
+  fun _ _ h => gi.dual.strict_mono_u h
+#align galois_coinsertion.strict_mono_l GaloisCoinsertion.strict_mono_l
+
+theorem isGLB_of_l_image [Preorder α] [Preorder β] (gi : GaloisCoinsertion l u) {s : Set α} {a : β}
+    (hs : IsGLB (l '' s) a) : IsGLB s (u a) :=
+  gi.dual.isLUB_of_u_image hs
+#align galois_coinsertion.is_glb_of_l_image GaloisCoinsertion.isGLB_of_l_image
+
+theorem isLUB_of_l_image [Preorder α] [Preorder β] (gi : GaloisCoinsertion l u) {s : Set α} {a : β}
+    (hs : IsLUB (l '' s) a) : IsLUB s (u a) :=
+  gi.dual.isGLB_of_u_image hs
+#align galois_coinsertion.is_lub_of_l_image GaloisCoinsertion.isLUB_of_l_image
+
+section lift
+
+variable [PartialOrder α]
+
+-- See note [reducible non instances]
+/-- Lift the infima along a Galois coinsertion -/
+@[reducible]
+def liftSemilatticeInf [SemilatticeInf β] (gi : GaloisCoinsertion l u) : SemilatticeInf α :=
+  { ‹PartialOrder α› with
```

Comment:
In this and the following few defs, the elaborator struggled with `αᵒᵈ` vs `α`; now it compiles but *much* slower than in mathlib3.

Reply:
I've left a porting note about this.